### ISSUE TITLE
Question re: DeepSDF training

### ISSUE THREAD

**Issue by extragoya** [marian42/shapegan#7]
Hi, thanks for your work and this repository! I was taking a look at the original DeepSDF training code, and your own training code for the autodecoder, and I noticed some differences. I was hoping I could perhaps pick your brain on your own experiences with DeepSDF-style training? 

For instance, I notice that your training completely shuffles the data samples across all points and across all shapes. So a batch will contain a random assortment of shapes. Whereas the DeepSDF code will instead randomly pick a set of shapes, and from those randomly select the points. So DeepSDF batches are only composed of a handful of shapes. Was there a reason you departed from the DeepSDF batch composition? I'm curious if you found it more stable? 

I also noticed that your autodecoder model includes no normalization layers. Was there a reason you decided not to include those? 

Because of these differences, I'm just wondering if there are any quick tips/strategies you can pass on - thanks in advance if you have the time to respond!

**Comment by marian42** [comment]
I'm afraid I don't have good answers for both of these questions.

I remember trying batch normalization in my DeepSDF implementation, but IIRC training was unsuccessful with that. (The DeepSDF authors say that batch normalization didn't work for them either, but weight normalization does work)

For the batch composition, I changed the implementation of the data loader throughout development and used the most "convenient" way of composing the batches. The current implementation expects all data to be in a single .to file, so this would of course fail if there is not enough RAM. I don't remember finding that the composition of the batch (few shapes vs many shapes) makes a difference, although testing for this parameter wasn't my goal.

If you want to be sure, you should test different batch compositions and BN layers.

**Comment by extragoya** [comment]
Thanks for taking the time to respond 

> If you want to be sure, you should test different batch compositions and BN layers.

Yeah, definitely - doing that now. Was just trying to bootstrap off of you :)

Thanks again for sharing your code - best of luck in your research.

**Comment by aluo-x** [comment]
Just to document a few things, totally unscientific.
I've also been experimenting with different normalizations on the generator w/ the point discriminator.

The following experiments were done with the same discriminator, and only the generator was modified. I train for only ~100 epochs, and visually check if the results look like they are converging.

* Spectral norm - fails, most values from the sdf network are very extreme
* Weak spectral norm (only apply norm if value is greater than 1) - fails, either invalid values or only shapes for cones/blobs
* Instance norm - works, but visually seems worse than layer/weight norm
* Group norm - works, I wonder if this would help in the multi-class case
* Weight norm - works (haven't benchmarked, visually looks comparable to layer norm, seems less diverse)
* Layer norm (repo default)

Seems quite robust to choice of activation:
* Leaky relu w/ slope 0.1 - works
* ELU - works
* ReLU (repo default)

Edit: training with instance norm is less stable than layer norm, at least in terms of generator loss.
